# Solving Systems of Linear Equations with NumPy - Lab

## Introduction 

Now you've gathered all the required skills needed to solve systems of linear equations. You saw why there was a need to calculate inverses of matrices, followed by matrix multiplication to figure out the values of unknown variables. 

The exercises in this lab present some problems that can be converted into a system of linear equations. 

## Objectives
You will be able to:

- Use matrix algebra and NumPy to solve a system of linear equations given a real-life example 
- Use NumPy's linear algebra solver to solve for systems of linear equations

## Exercise 1

A coffee shop is having a sale on coffee and tea. 

On day 1, 29 bags of coffee and 41 bags of tea were sold, for a total of 490 dollars.

On day 2, they sold 23 bags of coffee and 41 bags of tea, for which customers paid a total of 448 dollars.  

How much does each bag cost?

In [1]:
import numpy as np

In [9]:
# Create and solve the relevant system of equations
s = np.array([[29, 41], [23, 41]])
r = np.array([[490, 448]])

r = r.T
s_inv = np.linalg.inv(s)

ppb = s_inv.dot(r)
ppb

array([[7.],
       [7.]])

#### Describe your result
Each bag is worth $7

## Exercise 2

The cost of admission to a popular music concert was 162 dollars for 12 children and 3 adults. 

The admission was 122 dollars for 8 children and 3 adults in the same music concert. 

How much was the admission for each child and adult?

In [11]:
# Create and solve the relevant system of equations

c = np.array([[12, 3], [8, 3]])
t = np.array([[162, 122]])

t = t.T
c_inv = np.linalg.inv(c)

cost = c_inv.dot(t)
cost

array([[10.],
       [14.]])

#### Describe your result
Ten for the kids, 14 for the adults

## Exercise 3

You want to make a soup containing tomatoes, carrots, and onions.

Suppose you don't know the exact mix to put in, but you know there are 7 individual pieces of vegetables, and there are twice as many tomatoes as onions, and that the 7 pieces of vegetables cost 5.25 USD in total. 
You also know that onions cost 0.5 USD each, tomatoes cost 0.75 USD and carrots cost 1.25 USD each.

Create a system of equations to find out exactly how many of each of the vegetables are in your soup.

In [22]:
# Create and solve the relevant system of equations

c = np.array([[.75, .5, 1.25], [1, 1, 1], [1, -2, 0]])
t = np.array([[5.25, 7, 0]])

t = t.T
c_inv = np.linalg.inv(c)

count = c_inv.dot(t)
count

array([[4.],
       [2.],
       [1.]])

#### Describe your result
4 toms, 2 Os, 1 car

## Exercise 4

A landlord owns 3 properties: a 1-bedroom, a 2-bedroom, and a 3-bedroom house. 

The total rent he receives is 1240 USD. 

He needs to make some repairs, where those repairs cost 10% of the 1-bedroom house’s rent. The 2-bedroom repairs cost 20% of the 2-bedroom rental price and 30% of the 3-bedroom house's rent for its repairs.  The total repair bill for all three houses was 276 USD. 

The 3-bedroom house's rent is twice the 1-bedroom house’s rent. 

How much is the individual rent for three houses?

In [25]:
# Create and solve the relevant system of equations
a = np.array([[1, 1, 1], [.1, .2, .3], [2, 0, -1]])
b = np.array([[1240, 276, 0]])

b = b.T
a_inv = np.linalg.inv(a)

rents = a_inv.dot(b)
rents

array([[280.],
       [400.],
       [560.]])

#### Describe your result
one br - 280  
two br - 400  
three br - 560

## Summary
In this lab, you learned how to use NumPy to solve linear equations by taking inverses and matrix multiplication and also using numpy's `solve()` function. You'll now take these skills forward and see how you can define a simple regression problem using linear algebra and solve it with Numpy. 